### PR TITLE
Add :expandtabs and :noexpandtabs

### PR DESCRIPTION
**Pull request by AndrewLaneX**

### CODE REVIEW

File: babi/file.py [hunk 1]
```diff
@@ -524,9 +524,15 @@ def _indent_selection(self, margin: Margin) -> None:
...
+            tab = ' ' * self.buf.tab_size
+        else:
+            tab = '\t'
+
         for l_y in range(s_y, e_y + 1):
             if self.buf[l_y]:
-                self.buf[l_y] = ' ' * self.buf.tab_size + self.buf[l_y]
+                self.buf[l_y] = tab + self.buf[l_y]
                 if l_y == self.buf.y:
                     self.buf.x += self.buf.tab_size
```
Comment (asottile) [inline babi/file.py:537]: I assume this needs to be adjusted too

File: babi/file.py [hunk 2]
```diff
@@ -537,8 +543,16 @@ def _indent_selection(self, margin: Margin) -> None:
...
+        expandtabs = self.buf.expandtabs
+        if expandtabs:
+            tab = n * ' '
+        else:
+            tab = '\t'
+        self.buf[self.buf.y] = line[:self.buf.x] + tab + line[self.buf.x:]
+        if expandtabs:
+            self.buf.x += n
+        else:
+            self.buf.x += 1
```
Comment (asottile) [inline babi/file.py:555]: maybe make a

```
@property
def tab_string(self):
    ...
```

and then you can use the `len(...)` of that

File: babi/file.py [hunk 3]
```diff
@@ -524,20 +524,25 @@ def _indent_selection(self, margin: Margin) -> None:
...
     @edit_action('insert tab', final=False)
     def _tab(self, margin: Margin) -> None:
-        n = self.buf.tab_size - self.buf.x % self.buf.tab_size
+        tab_string = self.buf.tab_string
         line = self.buf[self.buf.y]
-        self.buf[self.buf.y] = line[:self.buf.x] + n * ' ' + line[self.buf.x:]
+        n = len(tab_string) - self.buf.x % len(tab_string)
+        self.buf[self.buf.y] = (
+            line[:self.buf.x] + tab_string + line[self.buf.x:]
+        )
```
Comment (asottile) [inline babi/file.py:545]: maybe I'm reading this wrong but I don't see how this still works correctly.

For example, if we're at position 1 and we press tab it should insert 3 spaces and not 4

Comment (asottile) [reply babi/file.py:545]: ah I see now, I'll improve the test: #93

Comment (asottile) [reply babi/file.py:545]: if you rebase it should show the test failure

File: babi/file.py [hunk 4]
```diff
@@ -524,20 +524,29 @@ def _indent_selection(self, margin: Margin) -> None:
...
         self.selection.set(sel_y, sel_x, self.buf.y, self.buf.x)
 
     @edit_action('insert tab', final=False)
     def _tab(self, margin: Margin) -> None:
         n = self.buf.tab_size - self.buf.x % self.buf.tab_size
+        tab_string = self.buf.tab_string
+        if tab_string != '\t':
+            tab_string = tab_string[:n]
+        else:
+            n = len(tab_string) - self.buf.x % len(tab_string)
```
Comment (asottile) [inline babi/file.py:545]: let's change this to:

```
if tab_string == '\t':
    n = 1
else:
    n = self.buf.tab_size - self.buf.x % self.buf.tab_size
    tab_string = tab_string[:n]
```

**Comment by AndrewLaneX** [comment]
> unindent does not work, will need to fix that!

Fixed :+1:

File: babi/file.py [hunk 5]
```diff
@@ -548,9 +557,12 @@ def tab(self, margin: Margin) -> None:
             self._tab(margin)
 
     def _dedent_line(self, s: str) -> int:
-        bound = min(len(s), self.buf.tab_size)
+        tab_size = 1
+        if self.buf.expandtabs:
+            tab_size = self.buf.tab_size
```
Comment (asottile) [inline babi/file.py:562]: isn't this just `tab_size = len(self.buf.tab_string)`

Comment (AndrewLaneX) [reply babi/file.py:562]: Oops, I missed that :thinking:

**Review by asottile — APPROVED** [review]

**Pull request merged**